File: day11.hs
import qualified Data.Map as Map

maxSteps :: Int = 75

type Cache = Map.Map Int Int

convert :: Int -> Int -> Int
convert c i = (c - 1) + i * maxSteps

countBlink :: Cache -> Int -> Int -> (Cache, Int)
countBlink cache 0 i = (cache, 1)
countBlink cache c 0 = countBlink cache (c - 1) 1
countBlink cache c i =
  case Map.lookup (convert c i) cache of
    Just r -> (cache, r)
    Nothing ->
      let s = show i in
      let l = length s in
      if even l then
        let (sl, sr) = splitAt (div l 2) s in
        let (nl, nr) = (read sl, read sr) in
        let (cache2, resl) = countBlink cache (c - 1) nl in
        let (cache3, resr) = countBlink cache2 (c - 1) nr in
        let res = resl + resr in
        let cache4 = Map.insert (convert c i) res cache3 in
        (cache4, res)
      else
        countBlink cache (c - 1) (i * 2024)

main :: IO ()
main = do
  input <- readFile "./input"
  let nums = read <$> words input
  let sol1 = sum $ snd . countBlink Map.empty 25 <$> nums
  putStrLn $ "First task: " ++ show sol1
  let sol2 = sum $ snd . countBlink Map.empty 75 <$> nums
  putStrLn $ "Second task: " ++ show sol2

File: day11.rs
use std::collections::HashMap;

static MAX_STEPS: usize = 75;
type Cache = HashMap<usize, usize>;

fn convert(c: usize, n: usize) -> usize {
    assert!(0 < c && c <= MAX_STEPS);
    (c - 1) + n * MAX_STEPS
}

fn count(cache: &mut Cache, c: usize, n: usize) -> usize {
    if c == 0 {
        1
    } else if n == 0 {
        count(cache, c - 1, 1)
    } else {
        match cache.get(&convert(c, n)) {
            Some(m) => *m,
            None => {
                let s = n.to_string();
                if s.len() % 2 == 0 {
                    let (sl, sr) = s.split_at(s.len() / 2);
                    let resl = count(cache, c - 1, sl.parse().unwrap());
                    let resr = count(cache, c - 1, sr.parse().unwrap());
                    let res = resl + resr;
                    cache.insert(convert(c, n), res);
                    res
                } else {
                    count(cache, c - 1, n * 2024)
                }
            }
        }
    }
}

fn main() {
    let mut cache = HashMap::new();
    let nums = [5910927, 0, 1, 47, 261223, 94788, 545, 7771];
    for (c, text) in [(25, "First"), (75, "Second")] {
        let mut r: usize = 0;
        for n in nums {
            r += count(&mut cache, c, n);
        }
        println!("{} task: {}", text, r);
    }
}
